pkg/noun/events.c

```diff
@@ -179,12 +179,47 @@ u3e_check(c3_c* cap_c)
 /* _ce_flaw_protect(): protect page after fault.
 */
 static inline c3_i
-_ce_flaw_protect(c3_w pag_w)
+_ce_flaw_protect(c3_w pag_w, c3_w gar_w)
 {
-  if ( 0 != mprotect(_ce_ptr(pag_w), _ce_page, (PROT_READ | PROT_WRITE)) ) {
-    fprintf(stderr, "loom: fault mprotect (%u): %s\r\n",
-                     pag_w, strerror(errno));
-    return 1;
+  // NB: must be static, since the stack is grown via page faults, and
+  // we're already in a page fault handler.
+  //
+  static c3_y con_y[16384];
+
+  if ( 0 == u3P.eph_i ) {
+    if ( 0 != mprotect(_ce_ptr(pag_w), _ce_page, (PROT_READ | PROT_WRITE)) ) {
+      fprintf(stderr, "loom: fault mprotect (%u): %s\r\n",
+                       pag_w, strerror(errno));
+      return 1;
+    }
+  }
+  else {
+    // save contents of page, to be restored after the mmap
+    //
+    // NB: don't copy guard page because unnecessary, and it's PROT_NONE.
+    //
+    if ( gar_w != pag_w ) {
+      memcpy(con_y, _ce_ptr(pag_w), _ce_page);
+    }
+
+    // map the dirty page into the ephemeral file
+    //
+    if ( MAP_FAILED == mmap(_ce_ptr(pag_w),
+                            _ce_page,
+                            (PROT_READ | PROT_WRITE),
+                            (MAP_FIXED | MAP_SHARED),
+                            u3P.eph_i, _ce_len(pag_w)) )
+    {
+      fprintf(stderr, "loom: fault mmap failed (%u): %s\r\n",
+                       pag_w, strerror(errno));
+      return 1;
+    }
+
+    // restore contents of page
+    //
+    if ( gar_w != pag_w ) {
+      memcpy(_ce_ptr(pag_w), con_y, _ce_page);
+    }
```

Comment:
`_ce_flaw_protect()` has two call sites, with three scenarios:

- u3e_ward(): retire the guard page (since we're moving it preemptively)
- u3e_fault():
  - retire the guard page (since we hit it)
  - dirty a persistent page

The old definition is sufficient for the guard page case; all that's need is to establish read-write protections. And the double-copy + remapping is only needed when dirtying a persistent page. So i think they should just be two different functions, with a little refactoring of the `u3e_fault()` call site (further distinguishing guard-page handling). That way, no pre-protection or guard-page logic is needed in the copy + remap function.

Reply:
Makes sense.  FYI, I'm out the rest of the week, so I'll do these tasks next week.

Reply:
Done